### PR TITLE
Feature/run simple cycle

### PR DESCRIPTION
**Pull request by tomalrussell**
Enables and tests running a two-model dependency cycle, with perfect convergence.

No notion here yet of a distance measure or proportion to assess near-convergence.

Nor any tests of more involved dependency structures, though they should "just work"^TM

### CODE REVIEW

File: smif/sos_model.py [hunk 1]
```diff
@@ -146,22 +148,104 @@ def _run_static_sos_model(self):
...
-
         """
         run_order = self._get_model_names_in_run_order()
         self.logger.info("Determined run order as %s", run_order)
         for timestep in self.timesteps:
-            for model_name in run_order:
-                logging.debug("Running %s for %d", model_name, timestep)
-                sector_model = self.model_list[model_name]
-                self._run_sector_model_timestep(sector_model, timestep)
+            for model_set in run_order:
```
Comment (willu47) [inline smif/sos_model.py]: Should a model_set be a class with methods `run` and `guess_results` and property `converged`?

File: smif/sos_model.py [hunk 2]
```diff
@@ -283,7 +367,8 @@ def _get_data(self, model, timestep):
 
                 elif source in self.model_list:
                     source_model = self.model_list[source]
-                    from_data = self.results[timestep][source][name]
+                    # get latest set of results from list
+                    from_data = self.results[timestep][source][-1][name]
```
Comment (willu47) [inline smif/sos_model.py]: `self.results` has an added dimension now for iteration?

File: smif/sos_model.py [hunk 3]
```diff
@@ -374,12 +464,34 @@ def _run_dynamic_optimisation(self):
         raise NotImplementedError
 
     def _get_model_names_in_run_order(self):
```
Comment (willu47) [inline smif/sos_model.py]: This is very neat.  I approve!

File: smif/sos_model.py [hunk 4]
```diff
@@ -414,7 +526,11 @@ def timesteps(self):
         list
             A list of timesteps
         """
-        return sorted(self._timesteps)
+        return self._timesteps
```
Comment (willu47) [inline smif/sos_model.py]: Add to the docstring above that the list is sorted?

File: tests/test_sos_model.py [hunk 5]
```diff
@@ -292,6 +305,72 @@ def test_timestep_before(self):
...
+        """
+        sos_model = get_sos_model_only_scenario_dependencies
+        ws_model = sos_model.model_list['water_supply']
+
+        results = sos_model.guess_results(ws_model, 2010)
+        sos_model._set_data(ws_model, 2010, results)
+
+        assert not sos_model.converged({'water_supply'}, 2010)
+
+    def test_converged_two_identical(self, get_sos_model_only_scenario_dependencies):
```
Comment (willu47) [inline tests/test_sos_model.py]: In the case where there are no interdependencies, is the convergence flag set upon completion of a successful simulation?  Does it need to be set?

File: smif/sos_model.py [hunk 6]
```diff
@@ -41,7 +41,9 @@ class SosModel(object):
 
     """
     def __init__(self):
+        # housekeeping
         self.logger = logging.getLogger(__name__)
+        self.max_iterations = 25
```
Comment (willu47) [inline smif/sos_model.py]: `max_iterations` Probably best set from an environment variable, or from a `ModelRun` configuration (`model.yaml` file)?

File: smif/sos_model.py (hunk 2 above)
Comment (tomalrussell) [inline smif/sos_model.py]: I've undone this, to hold iterations just within ModelSet.

File: smif/sos_model.py (hunk 1 above)
Comment (tomalrussell) [inline smif/sos_model.py]: Yes, though it ends up calling into SosModel quite a lot.. see what you think.

File: smif/sos_model.py (hunk 4 above)
Comment (tomalrussell) [inline smif/sos_model.py]: Done - https://github.com/nismod/smif/pull/66/commits/a8ea68529825a32381de1e9f6b3cf1868c9e395e

File: smif/sos_model.py (hunk 6 above)
Comment (tomalrussell) [inline smif/sos_model.py]: Yes, and defaults if not set.

File: smif/sos_model.py (hunk 3 above)
Comment (tomalrussell) [inline smif/sos_model.py]: Ta!

File: tests/test_sos_model.py (hunk 5 above)
Comment (tomalrussell) [inline tests/test_sos_model.py]: Not exactly a flag - convergence can be assessed of a ModelSet for a given timestep. If there is just a single model in a ModelSet, we take a shortcut past any iteration or tests for convergence.

File: smif/sos_model.py (hunk 1 above)
Comment (willu47) [inline smif/sos_model.py]: Hmm, I see what you mean. Without restructuring SosModel, we'll not solve this.

File: smif/sos_model.py [hunk 7]
```diff
@@ -601,6 +523,117 @@ def outputs(self):
         return parameter_model_map
 
 
+class ModelSet(object):
```
Comment (willu47) [inline smif/sos_model.py]: Better docstring here please ;)

**Pull request merged**